R/vpc-simulation.R

```diff
@@ -0,0 +1,480 @@
+    sim_cols = c("DV", "PRED"),
+    .join_col = c("NUM")
+){
+  check_model_object(.mod, c(NM_MOD_CLASS, NMSIM_MOD_CLASS))
+  checkmate::assert_numeric(n, lower = 1)
+  checkmate::assert_numeric(seed)
+
+  # Remove the following record types for simulation:
+  # $EST, $COV,  $TABLE, $SIMULATION
+  records_rm <- c("estimation", "covariance", "table", "simulation")
+  purrr::walk(records_rm, function(rec){
+    if(mod_has_record(.mod, type = rec)) remove_records(.mod, type = rec)
+  })
+
+  # Remove PK and prior records:
+  # $PRIOR, $THETA/$THETAP/$THETAPV, $OMEGA/$OMEGAP/$OMEGAPD, $SIGMA/$SIGMAP/$SIGMAPD
+  records_rm <- c("prior", "theta", "thetapv", "omega", "omegapd", "sigma", "sigmapd")
+  purrr::walk(records_rm, function(rec){
+    if(mod_has_record(.mod, type = rec)) remove_records(.mod, type = rec)
+  })
+
+
+  ## Add new $SIMULATION record (must be _after_ $ERROR or $PRED, but before $TABLE) ##
+  rep_name <- get_sim_replicate_record(.mod)
+  sim_lines <- glue("({seed}) SUBPROBLEMS={n} TRUE=FINAL")
+  add_new_record(.mod, "simulation", lines = sim_lines, after = rep_name)
+
+
+  ## Add new $TABLE record with predefined format and columns (including REPI) ##
+  table_name <- glue("{get_model_id(get_based_on(.mod)[1])}-sim.tab")
+  join_col_txt <- paste(.join_col, collapse = " ")
+  sim_col_txt <- paste(sim_cols, collapse = " ")
+  fmt_expr <- "FORMAT=s1PE12.5"
+  table_lines <- glue("ONEHEADER NOPRINT NOAPPEND {join_col_txt} {sim_col_txt} {fmt_expr} FILE={table_name}")
+  add_new_record(.mod, "table", lines = table_lines, after = "simulation")
+
+
+  ## Add new $MSFI record (end of file) ##
+  msf_path <- get_msf_path(read_model(get_based_on(.mod)[1]))
+  msf_path_rel <- fs::path_rel(msf_path, .mod[[ABS_MOD_PATH]])
+  msfi_lines <- glue("{msf_path_rel} NOMSFTEST")
+  add_new_record(.mod, "msfi", lines = msfi_lines, after = NULL)
+
+  return(invisible(.mod))
+}
+
+
+
+#' Check that a `bbr` model can be used for simulation
+#'
+#' Checks that a `bbr` model object is in the correct format to generate a
+#'  `bbi_nmsim_model` object.
+#' @inheritParams new_sim_model
+#' @returns `TRUE` invisibly
+#' @keywords internal
+check_model_for_sim <- function(.mod, .join_col){
+
+  # Check the .MSF file in $EST
+  msf_path <- get_msf_path(.mod, .check_exists = FALSE)
+  if(is.null(msf_path)){
+    # These symbols are not typical in NONMEM control streams
+    mod_id_sanit <- gsub("(-|_)", "", get_model_id(.mod))
+    rlang::abort(
+      c(
+        glue("`add_simulation` requires an MSF estimation record option."),
+        glue("e.g., `$EST MSFO={mod_id_sanit}.MSF`")
+      )
+    )
+  }
+
+  # Check .MSF file exists
+  msf_path_name <- basename(msf_path)
+  if(!fs::file_exists(msf_path)){
+    rlang::abort(glue("Could not find referenced MSF path ({msf_path_name}) at `{msf_path}`"))
+  }
+  # Note: file name must be all caps for now; otherwise it will be git ignored via bbi
+  if(msf_path_name != toupper(msf_path_name)){
+    rlang::abort(
```

Comment:
I'd rather not hardwire this into the workflow at this point. We should be able to do a quick release of `bbr` that let's `.msf` files stay around. I don't know who is using upper and lower case for the extension, but all the nonmem examples use `.msf` so it would be odd to error when using that extension.

Reply:
Im not sure how soon we'll be able to get to a `bbi` release (cc @seth127), but what would you prefer as the alternative? Would you just mention it in the docs? I feel like we cant pretend it's not an issue (i.e. have no handling or mention) for the first release of this feature anyways